errors with make, 'undefined reference'

Trying to get clickOS up and running, and hitting a brick wall when running the makefile for cosmos. I'm quite new to xen in general so I could just be hitting some really naive error and be unaware of it; if that's the case then please excuse my inexperience.

Here's the output of the makefile:

```
jeff@JeffSDNLaptop:~/cosmos$ sudo make
make: *** xen: No such file or directory.  Stop.
  LD  /home/jeff/cosmos/dist/bin/cosmos
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `xenstore_init':
clickos.c:(.text+0xc): undefined reference to `xs_daemon_open'
clickos.c:(.text+0x49): undefined reference to `xs_get_domain_path'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_read_handler':
clickos.c:(.text+0x135): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x161): undefined reference to `xs_write'
clickos.c:(.text+0x17d): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x19a): undefined reference to `xs_watch'
clickos.c:(.text+0x1cb): undefined reference to `xs_fileno'
clickos.c:(.text+0x204): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x226): undefined reference to `xs_read'
clickos.c:(.text+0x246): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x28d): undefined reference to `xs_unwatch'
clickos.c:(.text+0x2a2): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x2ce): undefined reference to `xs_write'
clickos.c:(.text+0x2ea): undefined reference to `xs_transaction_end'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_write_handler':
clickos.c:(.text+0x3a3): undefined reference to `xs_write'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_start':
clickos.c:(.text+0x5c8): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x636): undefined reference to `xs_read'
clickos.c:(.text+0x709): undefined reference to `xs_write'
clickos.c:(.text+0x732): undefined reference to `xs_write'
clickos.c:(.text+0x75b): undefined reference to `xs_set_permissions'
clickos.c:(.text+0x784): undefined reference to `xs_set_permissions'
clickos.c:(.text+0x7bb): undefined reference to `xs_write'
clickos.c:(.text+0x90c): undefined reference to `xs_write'
clickos.c:(.text+0x948): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x96f): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x99e): undefined reference to `xs_write'
clickos.c:(.text+0x9ba): undefined reference to `xs_transaction_end'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_stop':
clickos.c:(.text+0xa4b): undefined reference to `xs_transaction_start'
clickos.c:(.text+0xa77): undefined reference to `xs_write'
clickos.c:(.text+0xa93): undefined reference to `xs_transaction_end'
collect2: error: ld returned 1 exit status
make: *** [/home/jeff/cosmos/dist/bin/cosmos] Error 1
```

For reference, I've followed the instructions as found here: http://cnp.neclab.eu/getting-started#clickos

I've successfully built the image, and am currently at the step of 

```
For the toolstack, just clone the toolstack and run make.
```

Is this error the result of something like having mismatching versions of xen installed? ```xl info``` tells me I'm running 4.4.1, and I've downloaded the sources for 4.4.1.

Any help/advice would be appreciated!

Sure.

```
jeff@JeffSDNLaptop:~/cosmos$ sudo make distclean
make: *** xen: No such file or directory.  Stop.
  CLN /home/jeff/cosmos/build/libxcl/*.o
  CLN /home/jeff/cosmos/build/libcosmos/*.o
  CLN /home/jeff/cosmos/build/cosmos/*.o
  CLN /home/jeff/cosmos/build/libxcl
  CLN /home/jeff/cosmos/dist/lib/libxcl.so
  CLN /home/jeff/cosmos/build/libcosmos
  CLN /home/jeff/cosmos/dist/lib/libcosmos.so
  CLN /home/jeff/cosmos/dist/bin/cosmos
  CLN /home/jeff/cosmos/build
  CLN /home/jeff/cosmos/dist
```

And

```
jeff@JeffSDNLaptop:~/cosmos$ sudo make verbose=y
make: *** xen: No such file or directory.  Stop.
cc -std=gnu99 -fno-strict-aliasing -fPIC -Wall -Wstrict-prototypes -Wno-format-zero-length -lxenctrl -lxenguest -lxenstore -luuid -ldl -lutil  -o /home/jeff/cosmos/dist/bin/cosmos /home/jeff/cosmos/build/cosmos/main.o /home/jeff/cosmos/build/libcosmos/clickos.o /home/jeff/cosmos/build/libcosmos/domain_none.o
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `xenstore_init':
clickos.c:(.text+0xc): undefined reference to `xs_daemon_open'
clickos.c:(.text+0x49): undefined reference to `xs_get_domain_path'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_read_handler':
clickos.c:(.text+0x135): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x161): undefined reference to `xs_write'
clickos.c:(.text+0x17d): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x19a): undefined reference to `xs_watch'
clickos.c:(.text+0x1cb): undefined reference to `xs_fileno'
clickos.c:(.text+0x204): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x226): undefined reference to `xs_read'
clickos.c:(.text+0x246): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x28d): undefined reference to `xs_unwatch'
clickos.c:(.text+0x2a2): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x2ce): undefined reference to `xs_write'
clickos.c:(.text+0x2ea): undefined reference to `xs_transaction_end'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_write_handler':
clickos.c:(.text+0x3a3): undefined reference to `xs_write'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_start':
clickos.c:(.text+0x5c8): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x636): undefined reference to `xs_read'
clickos.c:(.text+0x709): undefined reference to `xs_write'
clickos.c:(.text+0x732): undefined reference to `xs_write'
clickos.c:(.text+0x75b): undefined reference to `xs_set_permissions'
clickos.c:(.text+0x784): undefined reference to `xs_set_permissions'
clickos.c:(.text+0x7bb): undefined reference to `xs_write'
clickos.c:(.text+0x90c): undefined reference to `xs_write'
clickos.c:(.text+0x948): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x96f): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x99e): undefined reference to `xs_write'
clickos.c:(.text+0x9ba): undefined reference to `xs_transaction_end'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_stop':
clickos.c:(.text+0xa4b): undefined reference to `xs_transaction_start'
clickos.c:(.text+0xa77): undefined reference to `xs_write'
clickos.c:(.text+0xa93): undefined reference to `xs_transaction_end'
collect2: error: ld returned 1 exit status
make: *** [/home/jeff/cosmos/dist/bin/cosmos] Error 1
```

```
jeff@JeffSDNLaptop:~/cosmos$ make verbose=y
xc: error: Could not obtain handle on privileged command interface (13 = Permission denied): Internal error
libxl: error: libxl.c:92:libxl_ctx_alloc: cannot open libxc handle: Permission denied
cannot init xl context
make: *** xen: No such file or directory.  Stop.
cc -std=gnu99 -fno-strict-aliasing -fPIC -Wall -Wstrict-prototypes -Wno-format-zero-length -lxenctrl -lxenguest -lxenstore -luuid -ldl -lutil  -o /home/jeff/cosmos/dist/bin/cosmos /home/jeff/cosmos/build/cosmos/main.o /home/jeff/cosmos/build/libcosmos/clickos.o /home/jeff/cosmos/build/libcosmos/domain_none.o
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `xenstore_init':
clickos.c:(.text+0xc): undefined reference to `xs_daemon_open'
clickos.c:(.text+0x49): undefined reference to `xs_get_domain_path'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_read_handler':
clickos.c:(.text+0x135): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x161): undefined reference to `xs_write'
clickos.c:(.text+0x17d): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x19a): undefined reference to `xs_watch'
clickos.c:(.text+0x1cb): undefined reference to `xs_fileno'
clickos.c:(.text+0x204): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x226): undefined reference to `xs_read'
clickos.c:(.text+0x246): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x28d): undefined reference to `xs_unwatch'
clickos.c:(.text+0x2a2): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x2ce): undefined reference to `xs_write'
clickos.c:(.text+0x2ea): undefined reference to `xs_transaction_end'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_write_handler':
clickos.c:(.text+0x3a3): undefined reference to `xs_write'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_start':
clickos.c:(.text+0x5c8): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x636): undefined reference to `xs_read'
clickos.c:(.text+0x709): undefined reference to `xs_write'
clickos.c:(.text+0x732): undefined reference to `xs_write'
clickos.c:(.text+0x75b): undefined reference to `xs_set_permissions'
clickos.c:(.text+0x784): undefined reference to `xs_set_permissions'
clickos.c:(.text+0x7bb): undefined reference to `xs_write'
clickos.c:(.text+0x90c): undefined reference to `xs_write'
clickos.c:(.text+0x948): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x96f): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x99e): undefined reference to `xs_write'
clickos.c:(.text+0x9ba): undefined reference to `xs_transaction_end'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_stop':
clickos.c:(.text+0xa4b): undefined reference to `xs_transaction_start'
clickos.c:(.text+0xa77): undefined reference to `xs_write'
clickos.c:(.text+0xa93): undefined reference to `xs_transaction_end'
collect2: error: ld returned 1 exit status
make: *** [/home/jeff/cosmos/dist/bin/cosmos] Error 1
```

```
jeff@JeffSDNLaptop:~/cosmos$ make verbose=y
xc: error: Could not obtain handle on privileged command interface (13 = Permission denied): Internal error
libxl: error: libxl.c:92:libxl_ctx_alloc: cannot open libxc handle: Permission denied
cannot init xl context
mkdir -p /home/jeff/cosmos/build
mkdir -p /home/jeff/cosmos/dist
mkdir -p /home/jeff/cosmos/dist/bin
mkdir -p /home/jeff/cosmos/dist/lib
mkdir -p /home/jeff/cosmos/build/libxcl
mkdir -p /home/jeff/cosmos/build/libcosmos
mkdir -p /home/jeff/cosmos/build/cosmos
cc -DXEN_VERSION=  -I/home/jeff/cosmos/include   -std=gnu99 -fno-strict-aliasing -fPIC -Wall -Wstrict-prototypes -Wno-format-zero-length  -c /home/jeff/cosmos/src/main.c -o /home/jeff/cosmos/build/cosmos/main.o
cc -DXEN_VERSION=  -I/home/jeff/cosmos/include   -std=gnu99 -fno-strict-aliasing -fPIC -Wall -Wstrict-prototypes -Wno-format-zero-length  -c /home/jeff/cosmos/src/clickos.c -o /home/jeff/cosmos/build/libcosmos/clickos.o
cc -DXEN_VERSION=  -I/home/jeff/cosmos/include   -std=gnu99 -fno-strict-aliasing -fPIC -Wall -Wstrict-prototypes -Wno-format-zero-length  -c /home/jeff/cosmos/src/domain_none.c -o /home/jeff/cosmos/build/libcosmos/domain_none.o
cc -std=gnu99 -fno-strict-aliasing -fPIC -Wall -Wstrict-prototypes -Wno-format-zero-length -lxenctrl -lxenguest -lxenstore -luuid -ldl -lutil  -o /home/jeff/cosmos/dist/bin/cosmos /home/jeff/cosmos/build/cosmos/main.o /home/jeff/cosmos/build/libcosmos/clickos.o /home/jeff/cosmos/build/libcosmos/domain_none.o
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `xenstore_init':
clickos.c:(.text+0xc): undefined reference to `xs_daemon_open'
clickos.c:(.text+0x49): undefined reference to `xs_get_domain_path'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_read_handler':
clickos.c:(.text+0x135): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x161): undefined reference to `xs_write'
clickos.c:(.text+0x17d): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x19a): undefined reference to `xs_watch'
clickos.c:(.text+0x1cb): undefined reference to `xs_fileno'
clickos.c:(.text+0x204): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x226): undefined reference to `xs_read'
clickos.c:(.text+0x246): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x28d): undefined reference to `xs_unwatch'
clickos.c:(.text+0x2a2): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x2ce): undefined reference to `xs_write'
clickos.c:(.text+0x2ea): undefined reference to `xs_transaction_end'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_write_handler':
clickos.c:(.text+0x3a3): undefined reference to `xs_write'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_start':
clickos.c:(.text+0x5c8): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x636): undefined reference to `xs_read'
clickos.c:(.text+0x709): undefined reference to `xs_write'
clickos.c:(.text+0x732): undefined reference to `xs_write'
clickos.c:(.text+0x75b): undefined reference to `xs_set_permissions'
clickos.c:(.text+0x784): undefined reference to `xs_set_permissions'
clickos.c:(.text+0x7bb): undefined reference to `xs_write'
clickos.c:(.text+0x90c): undefined reference to `xs_write'
clickos.c:(.text+0x948): undefined reference to `xs_transaction_end'
clickos.c:(.text+0x96f): undefined reference to `xs_transaction_start'
clickos.c:(.text+0x99e): undefined reference to `xs_write'
clickos.c:(.text+0x9ba): undefined reference to `xs_transaction_end'
/home/jeff/cosmos/build/libcosmos/clickos.o: In function `clickos_stop':
clickos.c:(.text+0xa4b): undefined reference to `xs_transaction_start'
clickos.c:(.text+0xa77): undefined reference to `xs_write'
clickos.c:(.text+0xa93): undefined reference to `xs_transaction_end'
collect2: error: ld returned 1 exit status
make: *** [/home/jeff/cosmos/dist/bin/cosmos] Error 1
```

On my system, in order to use xl commands I must sudo first. Otherwise I get the first few lines of error:
```
xc: error: Could not obtain handle on privileged command interface (13 = Permission denied): Internal error
libxl: error: libxl.c:92:libxl_ctx_alloc: cannot open libxc handle: Permission denied
cannot init xl context
```

For example:

```
jeff@JeffSDNLaptop:~/cosmos$ xl info
xc: error: Could not obtain handle on privileged command interface (13 = Permission denied): Internal error
libxl: error: libxl.c:92:libxl_ctx_alloc: cannot open libxc handle: Permission denied
cannot init xl context
```

--- a/Makefile
+++ b/Makefile
@@ -75,10 +75,10 @@ TOUCH			 = touch
 ASCOMPILE		 = $(CC) $(ASDEFINES) $(CDEFINES) $(CINCLUDES) $(ASFLAGS)
 COMPILE			 = $(CC) $(CDEFINES)  $(CINCLUDES) $(CPPFLAGS) $(CFLAGS)
 CCLD			 = $(CC)
-CCLINK			 = $(CCLD) $(CFLAGS) $(LDFLAGS) -o $@
+CCLINK			 = $(CCLD) $(CFLAGS) $(1) $(LDFLAGS) -o $@
 CXXCOMPILE		 = $(CXX) $(CDEFINES) $(CINCLUDES) $(CPPFLAGS) $(CXXFLAGS)
 CXXLD			 = $(CXX)
-CXXLINK			 = $(CXXLD) $(CXXFLAGS) $(LDFLAGS) -o $@
+CXXLINK			 = $(CXXLD) $(CXXFLAGS) $(1) $(LDFLAGS) -o $@
 
 SWIG			 = swig
 SWIGCOMPILE		 = $(SWIG) $(SWIGFLAGS) $(SWIGINCLUDES)
@@ -87,9 +87,9 @@ XXD				?= xxd
 ifneq ($(verbose),y)
 ascompile		 = @/bin/echo ' ' $(2) $< && $(ASCOMPILE) $(1)
 ccompile		 = @/bin/echo ' ' $(2) $< && $(COMPILE) $(1)
-cclink			 = @/bin/echo ' ' $(2) $@ && $(CCLINK) $(1)
+cclink			 = @/bin/echo ' ' $(2) $@ && $(CCLINK)
 cxxcompile		 = @/bin/echo ' ' $(2) $< && $(CXXCOMPILE) $(1)
-cxxlink			 = @/bin/echo ' ' $(2) $< && $(CXXLINK) $(1)
+cxxlink			 = @/bin/echo ' ' $(2) $< && $(CXXLINK)
 swig			 = @/bin/echo ' ' $(2) $< && $(SWIGCOMPILE) $(1)
 archive			 = @/bin/echo ' ' $(2) $@ && $(AR) cr $(1)
 x_verbose_cmd	 = $(if $(2),/bin/echo ' ' $(2) $(3) &&,) $(1) $(3)
@@ -97,9 +97,9 @@ verbose_cmd		 = @$(x_verbose_cmd)
 else
 ascompile		 = $(ASCOMPILE) $(1)
 ccompile		 = $(COMPILE) $(1)
-cclink			 = $(CCLINK) $(1)
+cclink			 = $(CCLINK)
 cxxcompile		 = $(CXXCOMPILE) $(1)
-cxxlink			 = $(CXXLINK) $(1)
+cxxlink			 = $(CXXLINK)
 swig			 = $(SWIGCOMPILE) $(1)
 archive			 = $(AR) crv $(1)
 x_verbose_cmd	 = $(1) $(3)

